src/grpc/infer_handler.h

```diff
@@ -652,12 +651,137 @@ class InferHandlerState {
       ctx_->set_compression_level(compression_level);
     }
 
+    void GrpcContextAsyncNotifyWhenDone(InferHandlerStateType* state)
+    {
+      ctx_->AsyncNotifyWhenDone(state);
+    }
+
+    bool IsCancelled() { return ctx_->IsCancelled(); }
+
     // Increments the ongoing request counter
     void IncrementRequestCounter() { ongoing_requests_++; }
 
     // Decrements the ongoing request counter
     void DecrementRequestCounter() { ongoing_requests_--; }
 
+
+    // Inserts the state to a set tracking active requests
+    // within the server core.
+    void InsertInflightState(InferHandlerStateType* state)
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      inflight_states_.insert(state);
+    }
+
+    // Erases the state to a set tracking active requests
+    // within the server core.
+    void EraseInflightState(InferHandlerStateType* state)
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      inflight_states_.erase(state);
+    }
+
+    // Issues the cancellation for all inflight requests
+    // being tracked by this context.
+    TRITONSERVER_Error* IssueRequestCancellation()
+    {
+      TRITONSERVER_Error* err = nullptr;
+      {
+        std::lock_guard<std::mutex> lock(mu_);
+
+        // Issues the request cancellation to the core.
+        for (auto state : inflight_states_) {
+          LOG_VERBOSE(1) << "Issuing cancellation for " << state->unique_id_;
+          bool is_cancelled = false;
+          err = TRITONSERVER_InferenceRequestIsCancelled(
+              state->irequest_ptr_, &is_cancelled);
```

Comment:
Do we need to bother checking? Or can we just cancel anyway? I don't think there's an error with "double" cancelling, just keeping the same bool true, but one less check to do.

Reply:
I am not sure. I was trying to conform with the API contract. For example, if `TRITONSERVER_InferenceRequestCancel` starts returning error for a request that was already cancelled, then we would break out of the loop early.

Reply:
I would tend to agree there is no need to check here. In terms of the public API there is not a way for the request to become cancelled except from the front end. (backend api doesn't have such an api to cancel the request).  I think we would always allow double cancellation. There is also no way to gurantee that between checking and cancelling that the request doesn't become cancelled (i.e. if in the future we allow for the backend to cancel). I think logging but ignoring the error to cancel is probably best.